File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_wishlist/lib/src/state/wishlist_state.dart
import '../models/wishlist_models.dart';

enum WishlistStatus { idle, loading, ready, error }

class WishlistState {
  const WishlistState({
    this.status = WishlistStatus.idle,
    this.lastResult,
    this.error,
  });

  final WishlistStatus status;
  final WishlistPageResult<Object?>? lastResult;
  final String? error;

  WishlistState loading() => const WishlistState(status: WishlistStatus.loading);

  WishlistState loaded(WishlistPageResult<Object?> result) =>
      WishlistState(status: WishlistStatus.ready, lastResult: result);

  WishlistState failed(String message) =>
      WishlistState(status: WishlistStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-wishlist/src/state/wishlistState.ts
import type { WishlistPageResult } from "../types/wishlistModels.js";

export interface WishlistState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: WishlistPageResult<unknown>;
  readonly error?: string;
}

export function initialWishlistState(): WishlistState {
  return { status: "idle" };
}

export function reduceWishlistState(
  state: WishlistState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: WishlistPageResult<unknown> }
    | { type: "failed"; error: string },
): WishlistState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
